spec/lang/representation.md

```diff
@@ -1,52 +1,57 @@
 # MiniRust representation relation
 
-The main purpose of [types](types.md) is to define how [values](values.md) are (de)serialized into/from memory.
-This is the *[representation relation]*, which is defined in the following.
-`decode` converts a list of bytes into a value; this operation can fail if the byte list is not a valid encoding for the given type.
-`encode` inverts `decode`; it will always work when the value is [well-formed][well-formed-value] for the given type (which the specification must ensure, i.e. violating this property is a spec bug).
-Unsized types cannot be represented as values and thus a call to `encode` or `decode` for such types can never occur in a well-formed program.
+The language specification enforces (by penalty of Undefined Behavior) that when a range of bytes is "used at a given type",
+it satisfies an invariant -- the *language invariant* (or "validity invariant").
+This is the main purpose of [types](types.md): to define how [values](values.md) are (de)serialized into/from memory.
+
+This *[representation relation]* for a type is defined in the following.
+`decode` converts a list of bytes into a (potentially not well-formed) value; this operation can already fail if the byte list is not a valid encoding for the given type.
+The values are always checked after decoding to be [well-formed][well-formed-value] by `check_value`, since some types also have additional runtime constraints.
+`encode` inverts `decode`; it will always succeed for [well-formed][well-formed-value] inputs.
+
+`encode`, `decode` and `check_value` satisfy some [properties](#generic-properties), which the specification must ensure, i.e. violating these properties is a spec bug.
+However, they can also make some assumptions which the specification ensures:
+The types these are called with/on are all [well-formed](well-formed.md#well-formed-layouts-and-types) and sized, as unsized types cannot be represented as values.
+`encode` can also assume the value is [well-formed][well-formed-value] for the type,
+and `decode` can assume the size of the byte list matches the size of the type.
```

Comment:
```suggestion
The representation relation is defined by two functions, which are duals to each other: `decode` and `encode`.
`decode` converts a list of bytes into a (potentially not well-formed) value; this operation can fail if the byte list is not a valid representation for the given type.
The decoded values must still be checked to be [well-formed][well-formed-value] by `check_value`, since some types also have additional runtime constraints.
`encode` inverts `decode`; it will always succeed for [well-formed][well-formed-value] inputs.

`encode`, `decode` and `check_value` satisfy some [properties](#generic-properties), which the specification must ensure, i.e. violating these properties is a spec bug.
However, they can also make some assumptions which the specification ensures:
The types these are called with/on are all [well-formed](well-formed.md#well-formed-layouts-and-types) and sized, as unsized types cannot be represented as values.
`encode` can also assume the value is [well-formed][well-formed-value] for the type,
and `decode` can assume the length of the byte list matches the size of the type.
```